File: tools/FindInsertLocation.java
package tools;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.text.Normalizer;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

public class FindInsertLocation {
    public static void main(String[] args) throws Exception {
        if (args.length == 0) {
            System.out.println("Usage: java FindInsertLocation <Drug Name> [drugs.json path]");
            System.exit(1);
        }

        String drugName = args[0];
        String jsonPath = args.length > 1 ? args[1] : "drugs.json";

        Path path = Paths.get(jsonPath);
        if (!Files.exists(path)) {
            System.err.println("drugs.json not found at: " + path.toAbsolutePath());
            System.exit(2);
        }

        List<String> lines = Files.readAllLines(path);

        // Collect top-level keys and their start line numbers
        List<KeyPos> keys = new ArrayList<>();
        for (int i = 0; i < lines.size(); i++) {
            String line = lines.get(i);
            String trimmed = line.trim();
            if (trimmed.startsWith("\"") && trimmed.contains("\":") && trimmed.endsWith("{")) {
                // match lines like:   "key": {
                int firstQuote = trimmed.indexOf('"');
                int secondQuote = trimmed.indexOf('"', firstQuote + 1);
                if (firstQuote >= 0 && secondQuote > firstQuote) {
                    String key = trimmed.substring(firstQuote + 1, secondQuote);
                    keys.add(new KeyPos(key, i + 1)); // 1-based line
                }
            }
        }

        if (keys.isEmpty()) {
            System.err.println("No top-level keys found in " + jsonPath);
            System.exit(3);
        }

        // Sort keys lexicographically to be safe (file may already be sorted)
        keys.sort(Comparator.comparing(k -> k.key));

        String slug = slugify(drugName);

        // Check if exists
        for (KeyPos kp : keys) {
            if (kp.key.equals(slug)) {
                System.out.println("Found existing key '" + kp.key + "' at line " + kp.line + " in " + jsonPath);
                System.exit(0);
            }
        }

        // Find insertion index
        int insertIndex = 0;
        while (insertIndex < keys.size() && keys.get(insertIndex).key.compareTo(slug) < 0) {
            insertIndex++;
        }

        KeyPos prev = insertIndex - 1 >= 0 ? keys.get(insertIndex - 1) : null;
        KeyPos next = insertIndex < keys.size() ? keys.get(insertIndex) : null;

        System.out.println("Suggested slug: '" + slug + "' (derived from '" + drugName + "')");
        if (prev != null) {
            System.out.println("Insert after: '" + prev.key + "' (line " + prev.line + ")");
        } else {
            System.out.println("Insert at start of file (before: '" + next.key + "' at line " + next.line + ")");
        }
        if (next != null) {
            System.out.println("Insert before: '" + next.key + "' (line " + next.line + ")");
        } else {
            System.out.println("Insert at end of file (after: '" + prev.key + "' at line " + prev.line + ")");
        }

        System.out.println();
        System.out.println("Suggested JSON snippet (drop this between the surrounding keys, include trailing comma if needed):");
        System.out.println();
        System.out.println(generateSnippet(slug, drugName));
    }

    static String generateSnippet(String slug, String prettyName) {
        StringBuilder sb = new StringBuilder();
        sb.append("  \"").append(slug).append("\": {\n");
        sb.append("    \"aliases\": [\"" + slug + "\"],\n");
        sb.append("    \"categories\": [\"stimulant\"],\n");
        sb.append("    \"name\": \"").append(slug).append("\",\n");
        sb.append("    \"pretty_name\": \"").append(prettyName).append("\",\n");
        sb.append("    \"properties\": {\n");
        sb.append("      \"summary\": \"Add a short summary here.\"\n");
        sb.append("    }\n");
        sb.append("  },");
        return sb.toString();
    }

    static String slugify(String input) {
        String nowhitespace = input.trim().toLowerCase();
        String normalized = Normalizer.normalize(nowhitespace, Normalizer.Form.NFKD);
        // remove diacritics
        normalized = normalized.replaceAll("\\p{M}", "");
        // replace non-alnum with hyphen
        normalized = normalized.replaceAll("[^a-z0-9]+", "-");
        // collapse hyphens
        normalized = normalized.replaceAll("-+", "-");
        // trim hyphens
        if (normalized.startsWith("-")) normalized = normalized.substring(1);
        if (normalized.endsWith("-")) normalized = normalized.substring(0, normalized.length() - 1);
        return normalized;
    }

    static class KeyPos {
        String key;
        int line;

        KeyPos(String key, int line) {
            this.key = key;
            this.line = line;
        }
    }
}

File: tools_2.0/tools/find_insert_location.py
import json
import sys
import os
import unicodedata
import re

def slugify(value):
    """
    Normalizes string, converts to lowercase, removes non-alpha characters,
    and converts spaces to hyphens.
    """
    value = str(value)
    value = unicodedata.normalize('NFKD', value).encode('ascii', 'ignore').decode('ascii')
    value = re.sub(r'[^\w\s-]', '', value).strip().lower()
    value = re.sub(r'[-\s]+', '-', value)
    return value

def generate_snippet(slug, pretty_name):
    snippet = {
        slug: {
            "aliases": [slug],
            "categories": [],
            "name": slug,
            "pretty_name": pretty_name,
            "properties": {
                "summary": "Add a short summary here."
            }
        }
    }
    # Convert to JSON string but remove the outer braces to make it a snippet
    json_str = json.dumps(snippet, indent=4)
    # Remove first and last lines (brackets)
    lines = json_str.split('\n')
    return '\n'.join(lines[1:-1])

def main():
    if len(sys.argv) < 2:
        print("Usage: python find_insert_location.py <Drug Name> [drugs.json path]")
        sys.exit(1)

    drug_name = sys.argv[1]
    
    # Determine path to drugs.json
    # Default to ../../drugs.json relative to this script
    base_dir = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
    default_json_path = os.path.join(base_dir, "drugs.json")
    
    json_path = sys.argv[2] if len(sys.argv) > 2 else default_json_path

    if not os.path.exists(json_path):
        print(f"Error: {json_path} not found.")
        sys.exit(2)

    try:
        with open(json_path, 'r', encoding='utf-8') as f:
            # We want to find line numbers, so we parse manually or use a library that preserves order/lines?
            # The java version read lines to find line numbers. 
            # Ideally we just want the keys in order. 
            # Let's verify we can find the keys.
            
            # Simple approach: Read file as lines to find line numbers for keys
            f.seek(0)
            lines = f.readlines()
            
        keys = []
        for i, line in enumerate(lines):
            line = line.strip()
            # Match "key": {  at top level roughly
             # Regex for "key": {
            match = re.match(r'^"([^"]+)":\s*\{', line)
            if match:
                keys.append({'key': match.group(1), 'line': i + 1})

        if not keys:
            print(f"No top-level keys found in {json_path}")
            sys.exit(3)

        keys.sort(key=lambda x: x['key'])

        slug = slugify(drug_name)

        # Check if exists
        for kp in keys:
            if kp['key'] == slug:
                print(f"Found existing key '{kp['key']}' at line {kp['line']} in {json_path}")
                sys.exit(0)

        # Find insertion index
        insert_index = 0
        while insert_index < len(keys) and keys[insert_index]['key'] < slug:
            insert_index += 1

        prev_key = keys[insert_index - 1] if insert_index - 1 >= 0 else None
        next_key = keys[insert_index] if insert_index < len(keys) else None

        print(f"Suggested slug: '{slug}' (derived from '{drug_name}')")
        if prev_key:
            print(f"Insert after: '{prev_key['key']}' (line {prev_key['line']})")
        else:
            print(f"Insert at start (before: '{next_key['key']}' at line {next_key['line']})")
        
        if next_key:
            print(f"Insert before: '{next_key['key']}' (line {next_key['line']})")
        else:
            print(f"Insert at end (after: '{prev_key['key']}' at line {prev_key['line']})")

        print("\nSuggested JSON snippet (add comma if needed):")
        print(generate_snippet(slug, drug_name))

    except Exception as e:
        print(f"Error processing file: {e}")
        sys.exit(1)

if __name__ == "__main__":
    main()
